src/lib/action.rs

```diff
@@ -0,0 +1,411 @@
+
+/// Apply an action to an AST node.
+pub trait ApplyAction<T: Clone, U: Clone> {
+    fn apply(&mut self, arena: &mut Arena<T, U>) -> Result<(), ArenaError>;
+}
+
+/// A list of actions to be applied.
+///
+/// This type will usually be used by iterating over the list and calling
+/// `apply()` on each element.
+pub type ActionList<T, U> = Vec<Box<ApplyAction<T, U>>>;
+
+/// Delete a node from a given AST.
+pub struct Delete {
+    node: NodeId,
+}
+
+/// Insert a new node into an AST as the `position`th child of an existing node.
+pub struct Insert<T: Clone, U: Clone> {
+    nth_child: u16,
+    data: T,
+    ty: U,
+    indent: u32,
+    new_parent: NodeId,
+}
+
+/// Move a node from one place to another within an AST.
+pub struct Move {
+    from_node: NodeId,
+    parent: NodeId,
+    pos: u16,
+}
+
+/// Update the data inside an AST node.
+pub struct Update<T: Clone, U: Clone> {
+    node: NodeId,
+    value: T,
+    ty: U,
+}
+
+impl fmt::Display for Delete {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "DEL {}", self.node)
+    }
+}
+
+impl<T: fmt::Display + Clone, U: fmt::Display + Clone> fmt::Display for Insert<T, U> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut indent = String::new();
+        for _ in 0..self.indent {
+            indent.push_str(" ");
+        }
+        write!(f,
+               "INS {}{} {} to {} at {}",
+               indent,
+               self.ty,
+               self.data,
+               self.new_parent,
+               self.nth_child)
+    }
+}
+
+impl fmt::Display for Move {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f,
+               "MOV {} to {} at {}",
+               self.from_node,
+               self.parent,
+               self.pos)
+    }
+}
+
+impl<T: fmt::Display + Clone, U: fmt::Display + Clone> fmt::Display for Update<T, U> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "UPD {} to {} {}", self.node, self.ty, self.value)
+    }
+}
+
+impl<T: Clone, U: Clone> ApplyAction<T, U> for Delete {
+    fn apply(&mut self, arena: &mut Arena<T, U>) -> Result<(), ArenaError> {
+        self.node.detach_with_children(arena)
+    }
+}
+
+impl<T: Clone, U: Clone> ApplyAction<T, U> for Insert<T, U> {
+    fn apply(&mut self, arena: &mut Arena<T, U>) -> Result<(), ArenaError> {
+        let mut new_id = arena.new_node(self.data.clone(), self.ty.clone(), self.indent);
+        new_id.make_nth_child_of(self.new_parent, self.nth_child, arena)
+    }
+}
+
+impl<T: Clone, U: Clone> ApplyAction<T, U> for Move {
+    fn apply(&mut self, arena: &mut Arena<T, U>) -> Result<(), ArenaError> {
+        self.from_node
+            .make_nth_child_of(self.parent, self.pos, arena)
+    }
+}
+
+impl<T: Clone, U: Clone> ApplyAction<T, U> for Update<T, U> {
+    fn apply(&mut self, arena: &mut Arena<T, U>) -> Result<(), ArenaError> {
+        if !arena.is_in_arena(self.node) {
```

Comment:
I think this was added in 679d2af0c19d46eab6455154b1723ae9c53e030f so I will change it in this PR.